openshift/console
# Bug 2064607: avoid pre-fetching tekton hub task versions in pipeline builder

**Fixes**: 

https://issues.redhat.com/browse/OCPBUGSM-42004

**Analysis / Root cause**: 

Prefetching all the task versions upfront might cause issues in slower network and this tasks catalog can grow in future.

**Solution Description**: 


1. Avoided prefetching of task versions and now fetching lazily when a task is selected.
2. Switching back and forth between previously selected tasks will not make API call again.


**Screen shots / Gifs for design review**: 

Pipeline builder network:

Before: 
<img width="1757" alt="image" src="https://user-images.githubusercontent.com/9964343/158871729-b3d31b57-fd9c-4ab2-969d-e347beb0afd9.png">

After:
![pipeline-builder-network-calls](https://user-images.githubusercontent.com/9964343/158870821-163cc3b1-7888-43bd-88b8-ed3c6326df63.png)

Video:

https://user-images.githubusercontent.com/9964343/158871491-ee678465-f98d-42f8-be0a-62d87d35926e.mov

**Unit test coverage report**: 

Fixed some existing tests and updated test data.

```
pipelineQuickSearchDetails:
    CTA button tests
      ✓ Add button should be disabled if the versions is not available (119ms)
      ✓ Add button should be enabled if the versions is not available (158ms)
      
 pipelineQuickSearchVersionDropdown:
    ✓ should not show version dropdown if an invalid value is passed as versions (18ms)
    ✓ should call the onchange handler with the version key (22ms)

  tektonHub.spec.ts:  
 getApiResponse:
    ✓ should throw error if the response status is not 200 (1ms)
    ✓ should throw error if the response ok is false (1ms)
    ✓ should return a valid response (1ms)
  
   getHubUIPath
    ✓ should return null if the path is not set
    ✓ should return a value if the path is set (1ms)
```
**Test setup:**

1. Install OpenShift Pipelines operator
2. Navigate to pipeline list page
3. Open the browser network inspector and filter for "api.hub.tekton.dev"
4. Click on Create Pipeline to navigate to pipeline builder page
5. Observe the network calls.



**Browser conformance**:
